File: LCA-binary-tree.py
'''
https://leetcode.com/problems/lowest-common-ancestor-of-a-binary-tree
Did it run on leetcode: Yes
Did you face any problem: Yes

Time Complexity: 0(N)
Space Complexity: 0(H) recursive stack

Algorithm:
- Do a post order traversal of the tree
- use bottom up approach when you have results for left and right node.
- And then compare if both left and right are not none then it's the LCA.

'''


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None


class Solution:
    def lowestCommonAncestor(self, root, p, q):
        """
        :type root: TreeNode
        :type p: TreeNode
        :type q: TreeNode
        :rtype: TreeNode
        """
        if(root == None):
            return None
        
        if( (root.val==p.val) or (root.val==q.val) ):
            return root
        
        left = self.lowestCommonAncestor(root.left,p,q)
        right = self.lowestCommonAncestor(root.right,p,q)
        
        if(left!=None and right!=None):
            return root
        
        if(left):
            return left
        elif(right):
            return right

File: LCABinaryTree.cpp
// Approach - DFS
// Time Complexity - O(n)
// Space Complexity -  O(2h)
// Problems Faced - It gives me Memory Limit Exceeded
// It runs on Leetcode - No - It gives me Memory Limit Exceeded.
class Solution {
    vector<TreeNode*> pathP;
    vector<TreeNode*> pathQ;
    private:
    void dfs(TreeNode* root, TreeNode* p, TreeNode* q, vector<TreeNode*> path){
        // base
        if(root == NULL)
            return;
        
        // logic
        //action
        path.push_back(root);
        if(root == p){
            pathP = path;
            pathP.push_back(p);
        }
        if(root == q){
            pathQ = path;
            pathQ.push_back(q);
        }
        // recursion
        dfs(root->left, p, q, path);
        dfs(root->right, p, q, path);
        
        // backtrack
        path.pop_back();
    }
public:
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        vector<TreeNode*> path;
        dfs(root, p, q, path);
        for(int i = 0; i < pathP.size(); i++){
            if(pathP[i] != pathQ[i])
                return pathP[i-1];
        }
        return NULL;
    }
};

// Approach - Bottom-Up Approach
// Time Complexity - O(n)
// Space Complexity - O(1)
// Problems Faced - No!
// It runs on Leetcode - Yes!
public:
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        if(root == NULL || root == p || root == q)
            return root;
        TreeNode* left = lowestCommonAncestor(root->left, p, q);
        TreeNode* right = lowestCommonAncestor(root->right, p, q);
        if(!left && !right)
            return NULL;
        else if(!left && right)
            return right;
        else if(!right && left)
            return left;
        else
            return root;
    }
};